modal/io_streams.py

```diff
@@ -356,6 +319,289 @@ async def aclose(self):
             await self._stream.aclose()
 
 
+async def _decode_bytes_stream_to_str(stream: AsyncGenerator[bytes, None]) -> AsyncGenerator[str, None]:
+    """Incrementally decode a bytes async generator as UTF-8 without breaking on chunk boundaries.
+
+    This function uses a streaming UTF-8 decoder so that multi-byte characters split across
+    chunks are handled correctly instead of raising ``UnicodeDecodeError``.
+    """
+    decoder = codecs.getincrementaldecoder("utf-8")(errors="strict")
+    async for item in stream:
+        text = decoder.decode(item, final=False)
+        if text:
+            yield text
+    # Flush any buffered partial character at end-of-stream
+    tail = decoder.decode(b"", final=True)
+    if tail:
+        yield tail
+
+
+async def _stream_by_line(stream: AsyncGenerator[bytes, None]) -> AsyncGenerator[bytes, None]:
+    """Yield complete lines only (ending with \n), buffering partial lines until complete."""
+    line_buffer = b""
+    async for message in stream:
+        assert isinstance(message, bytes)
+        line_buffer += message
+        while b"\n" in line_buffer:
+            line, line_buffer = line_buffer.split(b"\n", 1)
+            yield line + b"\n"
+
+    if line_buffer:
```

Comment:
nit: I prefer making explicit comparisons instead of relying on truthiness of non-boolean values, i.e. `if len(line_buffer) > 0` or `if line_buffer != b""`

To me this makes the types and intent of the code clearer to a reader, and (though not relevant in this case) avoids a certain set of bugs.

Reply:
I do think it's generally well-defined to do this for an array/container type so less bug-prone than doing it for something that's e.g. an optional array or something